hypershift-operator/controllers/nodepool/kubevirt/imagecacher.go

```diff
@@ -0,0 +1,238 @@
+
+	return nil
+}
+
+func (qi *cachedQCOWImage) cleanOldCaches(ctx context.Context, cl client.Client, oldDVs []v1beta1.DataVolume, logger logr.Logger) {
+	for _, oldDV := range oldDVs {
+		if oldDV.DeletionTimestamp == nil {
+			logger.Info("deleting an old boot image cache DataVolume", "namespace", oldDV.Namespace, "DataVolume name", oldDV.Name)
+			err := cl.Delete(ctx, &oldDV)
+			if err != nil {
+				logger.Error(err, fmt.Sprintf("failed to delete an old DataVolume; namespace: %s, name: %s", oldDV.Namespace, oldDV.Name))
+			}
+		}
+	}
+}
+
+func (qi *cachedQCOWImage) createDVForCache(ctx context.Context, cl client.Client, kvPlatform *hyperv1.KubevirtNodePoolPlatform) (*v1beta1.DataVolume, error) {
+	dv := qi.buildDVForCache(kvPlatform)
+
+	err := cl.Create(ctx, dv)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return nil, fmt.Errorf("failed to create a DataVolume for the boot image cache ; %w", err)
+	}
+
+	return dv, nil
+}
+
+func (qi *cachedQCOWImage) getDVSourceForVMTemplate() *v1beta1.DataVolumeSource {
+	return &v1beta1.DataVolumeSource{
+		PVC: &v1beta1.DataVolumeSourcePVC{
+			Namespace: qi.namespace,
+			Name:      qi.dvName,
+		},
+	}
+}
+
+func (qi *cachedQCOWImage) buildDVForCache(kvPlatform *hyperv1.KubevirtNodePoolPlatform) *v1beta1.DataVolume {
+	var (
+		volumeSize   apiresource.Quantity
+		storageClass *string
+	)
+
+	if kvPlatform.RootVolume != nil && kvPlatform.RootVolume.Persistent != nil {
+		if kvPlatform.RootVolume.Persistent.Size != nil {
+			volumeSize = *kvPlatform.RootVolume.Persistent.Size
+		}
+		storageClass = kvPlatform.RootVolume.Persistent.StorageClass
+	}
+
+	return &v1beta1.DataVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: bootImageNamePrefix,
+			Namespace:    qi.namespace,
+			Labels: map[string]string{
+				bootImageDVLabelRoleName: bootImageDVLabelRoleValue,
+			},
+			Annotations: map[string]string{
+				bootImageDVAnnotationHash:          qi.hash,
+				cdiImmediateBindingAnnotation:      "true",
+				cdiDeleteAfterCompletionAnnotation: "false",
+			},
+		},
+		Spec: v1beta1.DataVolumeSpec{
+			Source: &v1beta1.DataVolumeSource{
+				HTTP: &v1beta1.DataVolumeSourceHTTP{
+					URL: qi.name,
+				},
+			},
+			Preallocation: pointer.Bool(true),
+			Storage: &v1beta1.StorageSpec{
+				AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+				Resources: corev1.ResourceRequirements{
+					Requests: map[corev1.ResourceName]apiresource.Quantity{
+						corev1.ResourceStorage: volumeSize,
+					},
+				},
+				StorageClassName: storageClass,
+			},
+		},
+	}
+}
+
+var cacheDVSelector client.MatchingLabels = map[string]string{
+	bootImageDVLabelRoleName: bootImageDVLabelRoleValue,
+}
+
+func getCacheDVs(ctx context.Context, cl client.Client) ([]v1beta1.DataVolume, error) {
+	dvs := &v1beta1.DataVolumeList{}
+
+	err := cl.List(ctx, dvs, cacheDVSelector)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to read DataVolumes; %w", err)
+	}
+
+	return dvs.Items, nil
+}
```

Comment:
I think we need to have some sort of HostedCluster identifier within the cacheDVSelector in order to ensure we only select DVs associated with the cluster in question.

While at the moment, we have a 1:1 relationship of namespaces to HostedClusters, I don't think that assumption will hold true with external infra. It's possible that VMs for multiple HostedClusters and NodePools will exist in a single namespace in the future.

can you add the HostedCluster.Name as a label that we select on?

Reply:
makes a lot of sense. Will do